ui/src/app/shared/components/property-table/property-table.component.ts

```diff
@@ -0,0 +1,139 @@
+import { Component, Input, ViewChild } from '@angular/core';
+import { ModalDirective } from 'ngx-bootstrap';
+import { FormBuilder, FormControl, FormGroup } from '@angular/forms';
+import { Observable } from 'rxjs/Observable';
+import { Selectable } from '../../model/selectable';
+
+@Component({
+  selector: 'app-property-table',
+  templateUrl: './property-table.component.html'
+})
+export class PropertyTableComponent {
+
+  @Input() id: string;
+  @Input() titleText: string;
+  @Input() emptyText: string;
+  @Input() addText: string;
+
+  @ViewChild('childModal')
+  public childModal: ModalDirective;
+
+  private properties: Array<PropertyTableItem>;
+  singleForm: FormGroup;
+  bulkForm: FormGroup;
+  bulkProperties = new FormControl('', validateProperties);
+  singlePropertyKey = new FormControl('', validateKeyOrValue);
+  singlePropertyValue = new FormControl('', validateKeyOrValue);
+
+  constructor(fb: FormBuilder) {
+    this.properties = new Array();
+    this.singleForm = fb.group({
+      'singlePropertyKey': this.singlePropertyKey,
+      'singlePropertyValue': this.singlePropertyValue
+    });
+    this.bulkForm = fb.group({
+      'bulkProperties': this.bulkProperties
+    });
+  }
+
+  getProperties(): Array<PropertyTableItem> {
+    return this.properties;
+  }
+
+  getPropertiesAsObservable(): Observable<Array<PropertyTableItem>> {
+    return Observable.of(this.properties);
+  }
+
+  removeSelectedItems() {
+    this.properties = this.properties.filter(item => !item.isSelected);
+  }
+
+  addProperty() {
+    this.showChildModal();
+  }
+
+  public showChildModal(): void {
+    this.childModal.show();
+  }
+
+  public hideChildModal(): void {
+    this.childModal.hide();
+  }
+
+  cancel(): void {
+    this.hideChildModal();
+    this.clear();
+  }
+
+  clear(): void {
+    this.singlePropertyKey.setValue('');
+    this.singlePropertyValue.setValue('');
+    this.bulkProperties.setValue('');
+  }
+
+  submitBulkProperties(): void {
+    if (this.bulkProperties.value) {
+      for (const prop of this.bulkProperties.value.split('\n')) {
```

Comment:
hmm, trouble is that `validateProperties` will bail out with error for specific key when it finds something invalid. If we make this generic we'd need to return some sort of an error. Maybe we could leave this for further enhancement as it'd be nice to have a generic concept of how these similar things are done.